### ISSUE TITLE
Allow transitions in docstrings

### ISSUE THREAD

**Issue by gilbertohasnofb** [sphinx-doc/sphinx#7813]
**Is your feature request related to a problem? Please describe.**
I am trying to use transitions (`----`) in the docstrings of my Python package members but Sphinx outputs an error message when I do that and the generated html does not contain a transition.

Here is a minimal example. The function below belongs to `my_module`.

```
def foobar():
    r"""Foo

    ----

    Bar
    """
    pass
```

This is my Sphinx's main `index.rst` file:

```
.. autosummary::
    :toctree: _api_members

    my_module.foobar
```
Trying to make the html will output the following message:

`docstring of my_module.foobar:3: WARNING: Unexpected section title or transition.`

And as can be seen below, the output does not contain the expected transition.

![image](https://user-images.githubusercontent.com/5700795/84270754-143d7b00-ab23-11ea-89c0-260e52f2dc5f.png)

I believe the only changes I did to my default `conf.py` are:

```
extensions = [
    'sphinx.ext.autodoc',
    'sphinx.ext.autosummary',
]

autosummary_generate = True
autosummary_imported_members = True
```

**Describe the solution you'd like**
I would like to be able to use `----` in docstrings similarly to [regular rst](https://docutils.readthedocs.io/en/sphinx-docs/ref/rst/restructuredtext.html?#transitions).

**Describe alternatives you've considered**
The only way I found to solve this is to use raw html to print a `<hr>`, but this is really inelegant solution that polutes the docstrings.

```
def foobar():
    r"""Foo

    .. raw:: html

        <hr>

    Bar
    """
    pass
```

Producing the expected output with the horizontal rule:

![image](https://user-images.githubusercontent.com/5700795/84270831-36cf9400-ab23-11ea-9907-0b3754d03bc1.png)

**Comment by aaugustin** [comment]
With the Furo theme this is needed for proper rendering:

```rst
    .. raw:: html

        <hr class="docutils">
```

**Comment by tk0miya** [comment]
Unfortunately, it's a restriction of autodoc and autosummary. In docutils, transitions are prohibited inside the directives. For example, it can't be used inside `container` directive:

```
.. container::

   blah blah blah

   --------------

   blah blah blah
```

But, IMO, it would be better to remove such a restriction if possible. @shimizukawa What do you think?

**Comment by shimizukawa** [comment]
Agreed for "remove such a restriction" from docstrings because I don't feel confortable "docstring content is inside of a directive".
